traffic_ops/traffic_ops_golang/deliveryservice/servers/servers.go

```diff
@@ -263,7 +243,447 @@ func (dss *TODeliveryServiceServer) Delete(db *sqlx.DB, user auth.CurrentUser) (
 	rollbackTransaction = false
 	return nil, tc.NoError
 }
-func selectQuery() string {
+
+func selectQuery(orderBy string, limit string, offset string) (string, error) {
+
+	selectStmt := `SELECT
+	s.deliveryService,
+	s.server,
+	s.last_updated
+	FROM deliveryservice_server s`
+
+	allowedOrderByCols := map[string]string{
+		"":                "",
+		"deliveryservice": "s.deliveryService",
+		"server":          "s.server",
+		"lastupdated":     "s.last_updated",
+		"deliveryService": "s.deliveryService",
+		"lastUpdated":     "s.last_updated",
+		"last_updated":    "s.last_updated",
+	}
+	orderBy, ok := allowedOrderByCols[orderBy]
+	if !ok {
+		return "", errors.New("orderBy '" + orderBy + "' not permitted")
+	}
+
+	if orderBy != "" {
+		selectStmt += ` ORDER BY ` + orderBy
+	}
+
+	selectStmt += ` LIMIT ` + limit + ` OFFSET ` + offset + ` ROWS`
+	return selectStmt, nil
+}
+
+func deleteQuery() string {
+	query := `DELETE FROM deliveryservice_server
+	WHERE deliveryservice=:deliveryservice and server=:server`
+	return query
+}
+
+type DSServers struct {
+	DsId    *int  `json:"dsId" db:"deliveryservice"`
+	Servers []int `json:"servers"`
+	Replace *bool `json:"replace"`
+}
+
+type TODSServers DSServers
+
+
+func createServersForDsIdRef() *TODSServers {
+	var dsserversRef = TODSServers(DSServers{})
+	return &dsserversRef
+}
+
+func GetReplaceHandler(db *sqlx.DB) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		defer r.Body.Close()
+		handleErrs := tc.GetHandleErrorsFunc(w, r)
+		ctx := r.Context()
+		user, err := auth.GetCurrentUser(ctx)
+		if err != nil {
+			log.Errorf("unable to retrieve current user from context: %s", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		// get list of server Ids to insert
+		payload :=  createServersForDsIdRef() 
+
+		if err := json.NewDecoder(r.Body).Decode(payload); err != nil {
+			log.Errorf("Error trying to decode the request body: %s", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		servers := payload.Servers
+		dsId := payload.DsId
+
+		if servers == nil {
+			log.Error.Printf("no servers sent in POST; could not begin transaction: %v", err)
+			handleErrs(http.StatusBadRequest, err)
```

Comment:
This `err` is coming from above, `user, err := auth.GetCurrentUser(ctx)`, and will always be `nil`, but only by coincidence.

Can you change all 3 of these to `api.HandleErr(w, r, http.StatusBadRequest, errors.New("servers must exist"), nil)`, `"dsId must exist"`, `"replace must exist"`?

Incidentally, `handleErrs := tc.GetHandleErrorsFunc(w, r)` is a poor system (that's my fault), because it doesn't allow you to distinguish between server and user errors, and reply versus log appropriately. We need to move to `api.HandleErr`, which lets you distinguish and specify both in a single line.